### PR TITLE
Special char, 4 new regex, agt pool control, RTN flag,tkn cache, azsksettings framework revert, SVTResolver fixes

### PR DESCRIPTION
**Pull request by arvindsingh-cloud**
…settings revert, svtresolv changes

## Description
</br>
 Brief description of the requested change.

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1 [hunk 1]
```diff
@@ -172,7 +172,12 @@ class ADOSVTBase: SVTBase {
 								if ($null -ne $childResourceState.State.DataObject) {
 									if ($currentItem.StateManagement.CurrentStateData -and $null -ne $currentItem.StateManagement.CurrentStateData.DataObject) {
 										$currentStateDataObject = [JsonHelper]::ConvertToJsonCustom($currentItem.StateManagement.CurrentStateData.DataObject) | ConvertFrom-Json
-
+										try {
+										    $removeAccentCharectors = $this.RemoveAccentedCharacters(($currentStateDataObject | ConvertTo-Json -Depth 10))
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1:176]: Characters

File: src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1 [hunk 2]
```diff
@@ -172,7 +172,12 @@ class ADOSVTBase: SVTBase {
...
+										    $currentStateDataObject = ($removeAccentCharectors | ConvertFrom-Json)
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1:177]: characters

File: src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1 [hunk 3]
```diff
@@ -442,4 +447,17 @@ class ADOSVTBase: SVTBase {
 		
 		$this.UpdateControlStates($ControlResults);
 	}
+
+	# https://stackoverflow.com/questions/7836670/how-remove-accents-in-powershell
+	[string] RemoveAccentedCharacters($src) {
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1:452]: rename the variables

File: src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.AgentPool.ps1 [hunk 4]
```diff
@@ -62,14 +62,14 @@ class AgentPool: ADOSVTBase
             {
                   $agentPools = ($agentPoolsObj.fps.dataProviders.data."ms.vss-build-web.agent-pools-data-provider".taskAgentPools | Where-Object { ($_.autoProvision -eq $true -and $_.Name -eq $this.ResourceContext.resourcename) }) #| Select-Object @{Name = "Name"; Expression = {$_.Name}}
                   if (($agentPools | Measure-Object).Count -gt 0 ) {
-                    $controlResult.AddMessage([VerificationResult]::Passed,"Auto-provisioning is enabled for the $($agentPools) agent pools.");
+                    $controlResult.AddMessage([VerificationResult]::Failed,"Auto-provisioning is enabled for the $($agentPools) agent pool.");
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.AgentPool.ps1:65]: Measure-Object

File: src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.Build.ps1 [hunk 5]
```diff
@@ -85,28 +85,6 @@ class Build: ADOSVTBase
...
-                     }
-                     $ControlSettings = $null;
-                }
-                $resObj = $null;
-                }
-                catch {
-                    $controlResult.AddMessage($_);
-                }
-                
                 Get-Member -InputObject $this.BuildObj.variables -MemberType Properties | ForEach-Object {
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.Build.ps1:88]: logically split the operations of fetching variables and then looping

File: src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.Organization.ps1 [hunk 6]
```diff
@@ -229,7 +229,7 @@ class Organization: ADOSVTBase
             $stateData.Whitelisted_Extensions += $whiteListedExtensions
             $stateData.NonWhitelisted_Extensions += $NonwhiteListedExtensions
 
-            #$controlResult.SetStateData("Installed extensions list: ", $stateData);
+            $controlResult.SetStateData("Installed extensions list: ", $stateData);
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.Organization.ps1:232]: do for shared extension control as well

File: src/AzSK.AzureDevOps/Framework/Core/SVT/SVTControlAttestation.ps1 [hunk 7]
```diff
@@ -503,7 +503,16 @@ class SVTControlAttestation
 									else
 									{
 										$controlState = $this.ComputeEffectiveControlState($controlState, $controlSeverity, $isSubscriptionScan, $controlItem, $controlResult)										
-									}									
+									}
+									# keeping here to rectify only that perticular controls state data if any
+									try {
+										if ($controlState.State -and $controlState.State.DataObject) {
+											$removeAccentCharactors = $this.RemoveAccentedCharacters(($controlState.State.DataObject | ConvertTo-Json -Depth 10))
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/SVT/SVTControlAttestation.ps1:510]: characters

File: src/AzSK.AzureDevOps/Framework/Core/SVT/SVTControlAttestation.ps1 [hunk 8]
```diff
@@ -609,4 +618,16 @@ class SVTControlAttestation
 		
 		return $ValidAttestationStatesHashTable;
 	}
+
+	# https://stackoverflow.com/questions/7836670/how-remove-accents-in-powershell
+	[string] RemoveAccentedCharacters($src) {
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/SVT/SVTControlAttestation.ps1:623]: Move to helpers.ps1

File: src/AzSK.AzureDevOps/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 9]
```diff
@@ -67,20 +70,36 @@ class SVTResourceResolver: AzSKRoot {
                 throw [SuppressedException] "The parameter 'BuildNames' does not contain any string."
             }
         }
+        elseif ($ResourceTypeName -eq [ResourceTypeName]::Build -or $ResourceTypeName -eq [ResourceTypeName]::Build_Release) {
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/SVT/SVTResourceResolver.ps1:73]: B_R_S_A???

File: src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1 (hunk 1 above)
Comment (arvindsingh-cloud) [reply src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1:176]: changed

File: src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1 (hunk 2 above)
Comment (arvindsingh-cloud) [reply src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1:177]: changed

File: src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1 (hunk 3 above)
Comment (arvindsingh-cloud) [reply src/AzSK.AzureDevOps/Framework/Abstracts/ADOSVTBase.ps1:452]: renamed

File: src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.Organization.ps1 (hunk 6 above)
Comment (arvindsingh-cloud) [reply src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.Organization.ps1:232]: done..

File: src/AzSK.AzureDevOps/Framework/Core/SVT/SVTControlAttestation.ps1 (hunk 7 above)
Comment (arvindsingh-cloud) [reply src/AzSK.AzureDevOps/Framework/Core/SVT/SVTControlAttestation.ps1:510]: changed

File: src/AzSK.AzureDevOps/Framework/Core/SVT/SVTControlAttestation.ps1 (hunk 8 above)
Comment (arvindsingh-cloud) [reply src/AzSK.AzureDevOps/Framework/Core/SVT/SVTControlAttestation.ps1:623]: moved

File: src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.AgentPool.ps1 (hunk 4 above)
Comment (arvindsingh-cloud) [reply src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.AgentPool.ps1]: changed the api

File: src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.ServiceConnection.ps1 [hunk 10]
```diff
@@ -154,50 +154,28 @@ class ServiceConnection: ADOSVTBase
...
+
+                # fail the control if restricted group found on service connection
+                if($restrictedGroups)
                 {
-                    $controlResult.AddMessage([VerificationResult]::Failed,"Do not grant global security groups access to service connections. Granting elevated permissions to these groups can risk exposure of service connections to unwarranted individuals.");
-                    $controlResult.AddMessage("List of service connections granting access to global security groups:",$nonCompliantIdentities)
-                    $controlResult.SetStateData("List of service connections granting access to global security groups:",$nonCompliantIdentities)
+                    $controlResult.AddMessage([VerificationResult]::Failed,"Do not grant global groups access to service connections. Granting elevated permissions to these groups can risk exposure of service connections to unwarranted individuals.");
+                    $controlResult.AddMessage("Service connection granting access to global groups:",$restrictedGroups)
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.ServiceConnection.ps1:171]: Global groups that have access to the service connection

File: src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.ServiceConnection.ps1 [hunk 11]
```diff
@@ -154,50 +154,28 @@ class ServiceConnection: ADOSVTBase
...
+                    $controlResult.SetStateData("Service connection granting access to global groups:",$restrictedGroups)
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/SVT/AzureDevOps/AzureDevOps.ServiceConnection.ps1:172]: Global groups that have access to the service connection

File: src/AzSK.AzureDevOps/Framework/Core/SVT/SVTResourceResolver.ps1 [hunk 12]
```diff
@@ -41,7 +41,7 @@ class SVTResourceResolver: AzSKRoot {
             }
         }	
 
-        if ($ScanAllArtifacts) {
+        if ($ScanAllArtifacts -or ([string]::IsNullOrEmpty($ServiceConnectionNames) -and $ResourceTypeName -eq [ResourceTypeName]::ServiceConnection )) {
```
[the remark this answers is not in the document]
Comment (abhaydaga) [reply src/AzSK.AzureDevOps/Framework/Core/SVT/SVTResourceResolver.ps1:44]: Implicit behavior to scan all service connections if its list is not explicitly provided

**Pull request merged**